File: go/crawler/experiment.go
package crawler

import (
	"context"
	"fmt"
	"time"
	"webcrawler/utils"
)

func RunExperiment(crawler *WebCrawler, timeout time.Duration, maxConnections int) {
	startTime := time.Now()

	ctx, cancel := context.WithDeadline(context.Background(), startTime.Add(timeout))

	crawler.Crawl(ctx, utils.START_URL, maxConnections)
	cancel()

	elapsedTime := time.Since(startTime)

	fmt.Printf("explored=%d\n", len(crawler.SuccessfulExplored))
	fmt.Printf("found=%d\n", len(crawler.Found))
	fmt.Printf("totalChars=%d\n", crawler.CharsDownloaded)
	fmt.Printf("overheadTime=%d\n", (elapsedTime-timeout).Milliseconds())

	if utils.DEBUG {
		fmt.Println("Explored links:")
		for link := range crawler.SuccessfulExplored {
			fmt.Println(link)
		}

		fmt.Println("Found links:")
		for link := range crawler.Found {
			fmt.Println(link)
		}
	}
}

File: cats/shared/src/main/scala/Experiment.scala
package shared

import cats.effect._
import scala.concurrent.duration._
import java.sql.Time
import java.util.concurrent.TimeoutException
import scala.collection.immutable.HashSet

object Experiment {
  def run(
      crawler: WebCrawler,
      timeout: Int,
      maxConnections: Int,
      doCancel: Boolean
  ): IO[Unit] =
    for {

      successfulExploredR <- Ref.of[IO, HashSet[String]](HashSet.empty)
      foundR <- Ref.of[IO, HashSet[String]](HashSet.empty)
      charsDownloadedR <- Ref.of[IO, Int](0)
      _ <- IO(crawler.initRefs(foundR, successfulExploredR, charsDownloadedR))

      startTime = System.currentTimeMillis()
      fiber <- crawler
        .crawl(START_URL, maxConnections)
        .start
      // ! Need to control the timeout manually because the cancellation
      // does not work properly in Scala Native.
      // .timeout(FiniteDuration(timeout, "ms"))
      // .handleErrorWith(e =>
      //   e match
      //     case _: TimeoutException => IO(())
      //     case e: Throwable =>
      //       IO(println(s"Task failed with $e"))
      // )
      _ <- IO.sleep(FiniteDuration(timeout, "ms"))
      _ <- if doCancel then fiber.cancel else IO.unit
      endTime = System.currentTimeMillis()

      successfulExplored <- successfulExploredR.get
      found <- foundR.get
      charsDownloaded <- charsDownloadedR.get
    } yield
      println(s"explored=${successfulExplored.size}")
      println(s"found=${found.size}")
      println(s"totalChars=${charsDownloaded}")
      val elapsedTime = endTime - startTime
      println(s"overheadTime=${elapsedTime - timeout}")

      if DEBUG then
        println("Explored links:")
        successfulExplored.foreach(println(_))
        println("Found links:")
        found.foreach(println(_))
}
